### DISCUSSION TITLE
log4j2 extends abstractStringlayout to mask log no output traceId

### DISCUSSION THREAD

**Discussion by FnuJava** [apache/skywalking#11990] [Q&A]
my project extends abstractStringlayout  to mask log，when log4j print log Async then log no  output traceId 。

code：
public class CustomPatternLayout extends AbstractStringLayout {
    public static final Logger logger = LoggerFactory.getLogger(CustomPatternLayout.class);
    private PatternLayout patternLayout;

    protected CustomPatternLayout(Charset charset, String pattern) {
        super(charset);
        this.patternLayout = PatternLayout.newBuilder().withPattern(pattern).build();
    }

    public String hideMarkLog(String logStr) {
        try {
            if (!StringUtils.isBlank(logStr) && !CollUtil.isEmpty(Log4j2Rule.regulars)) {
                Iterator var2 = Log4j2Rule.regulars.iterator();

                while(var2.hasNext()) {
                    String regular = (String)var2.next();
                    if (ReUtil.count(regular, logStr) > 0) {
                        logStr = matchingAndEncrypt(logStr, regular);
                    }
                }

                return logStr;
            } else {
                return logStr;
            }
        } catch (Exception var4) {
            logger.info(">>>>>>>>> 脱敏处理异常 ERROR:{}", var4);
            return logStr;
        }
    }

    private static String matchingAndEncrypt(String msg, String regExp) {
        return ReUtil.replaceAll(msg, regExp, (matcher) -> {
            String group = matcher.group();
            if (org.apache.commons.lang.StringUtils.isNotEmpty(group)) {
                group = StrUtil.sub(group, 2, -2);
                return SecureUtil.sha256(group);
            } else {
                return msg;
            }
        });
    }

    @PluginFactory
    public static Layout createLayout(@PluginAttribute("pattern") String pattern, @PluginAttribute("charset") Charset charset) {
        return new CustomPatternLayout(charset, pattern);
    }

    public String toSerializable(LogEvent event) {
        return this.hideMarkLog(this.patternLayout.toSerializable(event));
    }
}

<?xml version="1.0" encoding="UTF-8"?>
 

 
<Configuration status="WARN">
 
    <Appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <CustomPatternLayout >
                <Pattern>%d %p %class{1.} [%t] %location %m %ex%n</Pattern>
            </CustomPatternLayout >
        </Console>
        
        <RandomAccessFile name="RandomAccessFile" fileName="./sync.log"
                          immediateFlush="false" append="false">
            <CustomPatternLayout >
                <Pattern>%d %p %class{1.} [%t] %location %m %ex%n</Pattern>
            </CustomPatternLayout >
        </RandomAccessFile>
    </Appenders>
    <Loggers>
        
        <AsyncLogger name="com.test" level="info" includeLocation="true" additivity="false">
            <AppenderRef ref="RandomAccessFile"/>
            <AppenderRef ref="Console"/>
        </AsyncLogger>
        <AsyncRoot level="info" includeLocation="true">
            <AppenderRef ref="RandomAccessFile"/>
        </AsyncRoot>
    </Loggers>
</Configuration>

**Comment by wu-sheng** [comment]
Async makes the trace context not available.
If you insist in async, you need to put the context into MDC, and read from it.

**Reply by FnuJava** [reply]
不是指的多线程异步，是log4j2开启异步打印日志，然后扩展abstractStringlayout实现日志脱敏就会不打印traceId。
如果单纯log4j2开启异步打印日志，没有扩展abstractStringlayout，是有traceId的`。`

**Reply by wu-sheng** [reply]
I am not going to follow CN conversation. If you need to be in Chinese, go to slack CN channel.

**Reply by FnuJava** [reply]
It does not refer to multi-threaded async。It happend when use **log4j2 Async** config and  extend Log4j abstractStringlayout class to Mask log，then no output traceId 。If you just open **log4j2 Async** config it can output traceId。

**Reply by wu-sheng** [reply]
That could be caused by some part of async log is enhanced by existing codes. But your new format is another thread from there.

Note, in Java, only thread matter for reading context(trace ID).

**Reply by FnuJava** [reply]
I fund in CustomPatternLayout class can not get traceId by TraceContext.TraceId() function。 Why Log4j2 default  async can get traceId？Do you have good idea to solve it？

**Reply by wu-sheng** [reply]
I have said, that this means it is not in one thread. You could verify that easily.
